packages/backend/src/open_payments/payment/incoming_remote/service.ts

```diff
@@ -57,114 +63,209 @@ async function create(
 ): Promise<
   OpenPaymentsIncomingPaymentWithPaymentMethods | RemoteIncomingPaymentError
 > {
-  const { walletAddressUrl } = args
-  const grantOrError = await getGrant(deps, walletAddressUrl, [
-    AccessAction.Create,
-    AccessAction.ReadAll
-  ])
-
-  if (isRemoteIncomingPaymentError(grantOrError)) {
-    return grantOrError
+  let walletAddress: OpenPaymentsWalletAddress
+
+  try {
+    walletAddress = await deps.openPaymentsClient.walletAddress.get({
+      url: args.walletAddressUrl
+    })
+  } catch (err) {
+    const errorMessage = 'Could not get wallet address'
+    deps.logger.error(
+      { err, walletAddressUrl: args.walletAddressUrl },
+      errorMessage
+    )
+    return RemoteIncomingPaymentError.UnknownWalletAddress
+  }
+
+  return createIncomingPayment(deps, {
+    createArgs: args,
+    walletAddress
+  })
+}
+
+async function createIncomingPayment(
+  deps: ServiceDependencies,
+  {
+    createArgs,
+    walletAddress,
+    retryOnTokenError = true
+  }: {
+    createArgs: CreateRemoteIncomingPaymentArgs
+    walletAddress: WalletAddress
+    retryOnTokenError?: boolean
+  }
+): Promise<
+  OpenPaymentsIncomingPaymentWithPaymentMethods | RemoteIncomingPaymentError
+> {
+  const resourceServerUrl =
+    walletAddress.resourceServer ?? new URL(walletAddress.id).origin
+
+  const grantOptions = {
+    authServer: walletAddress.authServer,
+    accessType: AccessType.IncomingPayment,
+    accessActions: [AccessAction.Create, AccessAction.ReadAll]
+  }
+
+  const grant = await deps.grantService.getOrCreate(grantOptions)
+
+  if (isGrantError(grant)) {
+    return RemoteIncomingPaymentError.InvalidGrant
   }
 
   try {
-    const url = new URL(walletAddressUrl)
     return await deps.openPaymentsClient.incomingPayment.create(
       {
-        url: url.origin,
-        accessToken: grantOrError.accessToken
+        url: resourceServerUrl,
+        accessToken: grant.accessToken
       },
       {
-        walletAddress: walletAddressUrl,
-        incomingAmount: args.incomingAmount
-          ? serializeAmount(args.incomingAmount)
+        walletAddress: walletAddress.id,
+        incomingAmount: createArgs.incomingAmount
+          ? serializeAmount(createArgs.incomingAmount)
           : undefined,
-        expiresAt: args.expiresAt?.toISOString(),
-        metadata: args.metadata ?? undefined
+        expiresAt: createArgs.expiresAt?.toISOString(),
+        metadata: createArgs.metadata ?? undefined
       }
     )
   } catch (err) {
     const errorMessage = 'Error creating remote incoming payment'
-    deps.logger.error({ err, walletAddressUrl }, errorMessage)
+
+    if (err instanceof OpenPaymentsClientError) {
+      if ((err.status === 401 || err.status === 403) && retryOnTokenError) {
+        await deps.grantService.delete(grant.id) // force new grant creation
+
+        return createIncomingPayment(deps, {
+          createArgs,
+          walletAddress,
+          retryOnTokenError: false
```

Comment:
we could make it such that we retry multiple times, but for our use case, we just want to retry with a new grant once